### DISCUSSION TITLE
define move semantics

### DISCUSSION THREAD

**Discussion by maddanio** [vlang/v#11915] [Ideas]
hello

so I am not sure if you are familiar with move semantics in c++. basically its very nice to have types (i.e. structs in v) wrap and manage resources. for example a file handle, which should be closed automatically. in order for this to work properly such resource handles have to be move-only types, i.e. cannot be copied, and only really destructed once. while it seems v effectively does moves wherever it can (the free of my test struct is only called once, even though I return the instance from a function) I cannot find that being properly defined and guarantueed, or am I overlooking something?

**Comment by Cons-Cat** [comment]
V is based around copy-on-write, not move semantics, similar to Go. Fwiw, I'm a much bigger fan of move semantics than CoW.

**Reply by JalonSolov** [reply]
CoW is more efficient for both memory usage _and_ performance, _unless_ you write.  Then it works about the same.

**Reply by Cons-Cat** [reply]
That is the untrue. CoW containers require a flag for whether it has been copied yet or not. That makes it consume more memory. 

Mutable CoW references require checking that flag on read access, that makes it less efficient. 

Move semantics allow a user to express through the type system whether a deep or shallow copy is required, so it is more efficient in both ways. Even for CoW containers in C++, move semantics can improve the case where you pass to a function and you know it will not copy and write.

**Reply by JalonSolov** [reply]
The struct may require a flag, but that is usually less space than copying the entire thing referred to.  After it's copied, it may require slightly more memory, but by that point you've already used up so much, one small extra field is not a "noticeable" amount.

**Reply by Cons-Cat** [reply]
A move is not supposed to be a deep copy. I feel like you may misunderstand what move semantics means in C++.

**Reply by maddanio** [reply]
And how do you copy a resource like a file?
